File: examples/zero_copy/talker.cc
// Copyright 2016 Open Source Robotics Foundation, Inc.
// Copyright 2022 wayve.ai
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <cstring>
#include <memory>

#include "chatter_interface/msg/chatter.hpp"
#include "rclcpp/rclcpp.hpp"

class MinimalPublisher : public rclcpp::Node {
 public:
  MinimalPublisher() : Node("minimal_publisher"), count_(0) {
    declare_parameter("callback_period_ms", 500);
    auto callback_period_ms = get_parameter("callback_period_ms").as_int();

    publisher_ = create_publisher<chatter_interface::msg::Chatter>("topic", 10);
    auto timer_callback = [this]() -> void {
      auto message = publisher_->borrow_loaned_message();
      std::string str = "Hello, world! " + std::to_string(count_++);
      auto copy_size = str.size() < message.get().data.size()
                           ? str.size()
                           : message.get().data.size();
      std::memcpy(message.get().data.begin(), str.c_str(), copy_size);
      message.get().data_length = copy_size;
      message.get().timestamp =
          std::chrono::duration_cast<std::chrono::microseconds>(
              std::chrono::steady_clock::now().time_since_epoch())
              .count();
      message.get().count = count_;
      RCLCPP_INFO(get_logger(), "Publishing: '%s'", str.c_str());
      publisher_->publish(std::move(message));
    };
    timer_ = create_wall_timer(std::chrono::milliseconds(callback_period_ms),
                               timer_callback);
  }

 private:
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<chatter_interface::msg::Chatter>::SharedPtr publisher_;
  size_t count_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MinimalPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: examples/zero_copy/talker.rs
use std::env;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::SystemTime;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let shut_down = Arc::new(AtomicBool::new(false));
    for signal in [signal_hook::consts::SIGINT, signal_hook::consts::SIGTERM] {
        signal_hook::flag::register(signal, Arc::clone(&shut_down))?;
    }

    let context = rclrs::Context::new(env::args())?;

    let node = rclrs::create_node(&context, "minimal_publisher")?;
    let callback_period_ms = node
        .declare_parameter::<i64>("callback_period_ms")
        .default(500)
        .read_only()
        .unwrap()
        .get();

    let publisher = node.create_publisher::<chatter_interface::msg::rmw::Chatter>(
        "topic",
        rclrs::QOS_PROFILE_DEFAULT,
    )?;

    let mut publish_count: u32 = 0;

    while !shut_down.load(Ordering::Relaxed) && context.ok() {
        let mut message = publisher.borrow_loaned_message()?;
        message.timestamp = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)?
            .as_micros() as u64;
        message.count = publish_count as u64;

        let msg = format!("Hello, world! {publish_count}");
        let msg_len = msg.len();
        message.data_length = msg_len as u64;
        message.data[..msg_len].copy_from_slice(msg.as_bytes());
        rclrs::log_info!(node.logger_name(), "Publishing: {}", msg);
        message.publish()?;
        publish_count += 1;
        std::thread::sleep(std::time::Duration::from_millis(callback_period_ms as u64));
    }
    Ok(())
}
